spec/index.bs

```diff
@@ -1191,131 +1209,176 @@ access to it before garbage collection by calling {{GPUBuffer/destroy()}}.
-    1. Enqueue an operation on the [=Queue timeline=] that will execute the following:
+      Given a {{GPUBuffer}} |this|, a {{GPUSize64}} |offset| and a {{GPUSize64}} |size|
+      the following validation rules apply:
 
-        1. Let |m| be a new {{ArrayBuffer}} of size the {{[[size]]}} of |this| that is filled with zeroes.
-        1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapped for writing=].
-        1. If |p| is pending:
+      1. |this|.{{[[state]]}} must be [=buffer state/mapped=] or [=buffer state/mapped at creation=].
+      1. |offset| must more or equal to |this|.{{[[mapping_range]]}}[0]
+      1. |offset| + |size| must be less or equal to |this|.{{[[mapping_range]]}}[0] + |this|.{{[[mapping_range]]}}[1]
 
-            1. Resolve |p| with |m|.
+      Note: It is valid to get mapped ranges of an error {{GPUBuffer}} that is [=buffer state/mapped at creation=] because
+      the [=Content timeline=] might not know it is an error {{GPUBuffer}}.
 
-    1. Return |p|.
+    </dfn>
+  </div>
 </div>
 
-### <dfn method for=GPUBuffer>unmap()</dfn> ### {#buffer-unmap}
+### <dfn method for=GPUBuffer>unmap()</dfn> ### {#GPUBuffer-unmap}
 
 <div algorithm="GPUBuffer.unmap">
 
-    1. If the {{[[state]]}} slot of |this| is [=buffer state/unmapped=] or [=buffer state/destroyed=]:
+  **Arguments:** (None)
 
-        1. Record a validation error on the current scope.
-        1. Return.
+  **Returns:** (None)
 
-    1. If the {{[[mapping]]}} slot of |this| is a {{Promise}}:
+  1. If this call doesn't follow the [$unmap Valid Usage$]:
 
-        1. [=Reject=] {{[[mapping]]}} with an {{AbortError}}.
-        1. Set the {{[[mapping]]}} slot of |this| to null.
+    1. Record a validation error on the current scope.
+    1. Return.
 
-    1. If the {{[[mapping]]}} slot of |this| is an {{ArrayBuffer}}:
+  1. If |this|.{{[[state]]}} is [=buffer state/mapping pending=]:
 
-        1. If the {{[[state]]}} slot of this is [=buffer state/mapped for writing=]:
+    1. [=Reject=] {{[[mapping]]}} with an {{AbortError}}.
+    1. Set the {{[[mapping]]}} slot of |this| to null.
 
-            1. Enqueue an operation on the [=Queue timeline=] that updates |this|'s allocation to the content of the {{ArrayBuffer}} in the {{[[mapping]]}} slot of |this|.
+  1. If |this|.{{[[state]]}} is [=buffer state/mapped=] or [=buffer state/mapped at creation=]:
 
-        1. Detach |this|.{{[[mapping]]}} from its content.
-        1. Set the {{[[mapping]]}} slot of |this| to null.
+    1. If one of the two following conditions holds:
 
-    1. Set the {{[[state]]}} slot of |this| to [=buffer state/unmapped=].
+        1. |this|.{{[[state]]}} is [=buffer state/mapped at creation=]
+        1. |this|.{{[[state]]}} is [=buffer state/mapped=] and |this|.{{[[usage]]}} contains {{GPUBufferUsage/MAP_WRITE}}
 
-</div>
+    1. Then:
+
+        1. Enqueue an operation on the [=Queue timeline=] that updates the |this|.{{[[mapping_range]]}}
```

Comment:
similarly, which queue?

Reply:
Right now the default queue. Otherwise it will depend on if/what multi-queue looks like. If the resource was created to be used on queues explicitly, you might need to do an ownership transfer before this is visible. IDK